Implement asyncmap/pmap over `map`

amitmurthy opened the pull request:
This is a complete rework of `asyncmap` as a twice `map` on the collection in order to leverage various specializations on `map` for `asyncmap` and `pmap`.

See https://github.com/JuliaLang/julia/issues/19126, https://github.com/JuliaLang/julia/issues/4606 and https://github.com/JuliaLang/julia/issues/14265

```
# currently works as expected for

asyncmap(identity, ones(2,2))
asyncmap(identity, sparse(eye(3)))
asyncmap(identity, (1,2,3))

pmap(identity, ones(2,2))
pmap(identity, sparse(eye(3)))
pmap(identity, (1,2,3))

# WIP for:
asyncmap((x,y,z)->x+y+z, 1,2,3)
asyncmap(x->x, "Hello")
asyncmap(x->x, BitArray(4))
```

This implementation is also a bit faster compared to 0.5
```
julia> @time asyncmap(identity, 1:10^6);
  3.438046 seconds (9.11 M allocations: 264.920 MB, 6.19% gc time)

julia> @time asyncmap(identity, 1:10^6);
  3.399210 seconds (9.00 M allocations: 259.771 MB, 9.18% gc time)
```

0.5:
```
julia> @time asyncmap(identity, 1:10^6);
  3.919530 seconds (34.11 M allocations: 820.279 MB, 25.20% gc time)

julia> @time asyncmap(identity, 1:10^6);
  4.206910 seconds (33.99 M allocations: 815.587 MB, 28.50% gc time)
```

amitmurthy committed:
Implement asyncmap/pmap over `map`. Export asyncmap!

amitmurthy commented:
Ready for a review.

tkelman commented on base/asyncmap.jl:
````diff
@@ -3,161 +3,336 @@
+function is executed for different invocations of `ntasks` and `batch_size`.
+
+With ntasks undefined, each element is processed in a different task.
+```
````
should this be a doctest?

amitmurthy replied:
I considered that. Since the examples are printing out `object_id`s, wouldn't the doctest fail?

kshyatt commented:
What's the status on this?

amitmurthy commented:
Ready to be merged. Just leaving it open for a couple more days for feedback.

shashi commented:
> See #19126, #4606 and #4606

Did you mean to link to a third issue?

amitmurthy commented:
Yes - https://github.com/JuliaLang/julia/issues/14265

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+
+`ntasks` specifies the number of tasks to run concurrently.
+Depending on the length of the collections, if `ntasks` is unspecified,
+upto a 100 tasks will be used for concurrent mapping.
```
"up to" is 2 words

tkelman commented on base/asyncmap.jl:
````diff
@@ -3,161 +3,336 @@
+5
+```
+
+With `ntasks=2` all elements are only processed in 2 tasks.
````
"only" is unneeded here

tkelman commented on base/asyncmap.jl:
````diff
@@ -3,161 +3,336 @@
+```
+
+With `batch_size` defined, the mapping function needs to be changed to accept an array
+of argument tuples. And return an array of results. In this example we call `map`
````
"tuples, and return" otherwise this isn't a sentence

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+
+With `batch_size` defined, the mapping function needs to be changed to accept an array
+of argument tuples. And return an array of results. In this example we call `map`
+on the batch input to achieve the same.
```
not at all clear what "the same" is referring to

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+    ntasks = verify_ntasks(c[1], ntasks)
+
+    if isa(batch_size, Number)
+        @assert batch_size >= 0
```
should give an actual error message here

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+
+    if isa(ntasks, Function)
+        map_f = (x...) -> begin
+            # check number of tasks everytime, and start one if required.
```
"every time" is 2 words

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
-                itr.results[i] = itr.f(args...)
-                notify(state.item_done, nothing)
+    # check if there was a genuine problem with asyncrun
+    (asyncrun_excp != nothing) && throw(asyncrun_excp)
```
use `===` and `!==` to compare to `nothing`

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+        return asyncrun.x
+    else
+        # second run, extract values from the Refs and return
+        return map(x->x.x, asyncrun)
```
use a different variable name when the field is `.x` ?

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
-    state.active_count += 1
-    t
+    # Use an unbuffered channel for communicating with the worker tasks. In the event
+    # of any error in any of the worker tasks, the channel is closed. And this
```
"And" should either be continuing the previous sentence or removed

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
-    for _ in 1:itr.max_tasks()
-        start_collector_task(itr, state)
-        state.nfree += 1
+# map on a single BitArray returns a BitArray of the mapping function is boolean.
```
if the mapping function

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+collects output into results.
+
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+`asyncmap()`](:func:`asyncmap`). If `batch_size` is specified, `f` must
```
missing open [

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+be a function which operates on an array of argument tuples.
+
+!!! note
+    `next(::AsyncCollector, state) -> (nothing, state)`
```
explain why this is significant

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
+        foreach(x->(v=wait(x); isa(v, Exception) && throw(v)), state.worker_tasks)
+        empty!(state.worker_tasks)
+        return true
+     else
```
indentation is off

tkelman replied:
not addressed

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,336 @@
-Results are returned by the iterator as they become available.
+
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+`asyncmap()`](:func:`asyncmap`). If `batch_size` is specified, `f` must
```
missing [

tkelman commented on base/asyncmap.jl:
```diff
@@ -220,26 +384,22 @@ size(itr::AsyncGenerator) = size(itr.collector.enumerator)
-asyncmap!(f, c) = (for x in AsyncCollector(f, c, c) end; c)
-
+function asyncmap!(f, c; ntasks=0, batch_size=nothing)
+    foreach(_->_, AsyncCollector(f, c, c; ntasks=ntasks, batch_size=batch_size))
```
identity is easier to read than `_->_`

amitmurthy commented:
Incorporated feedback. Meanwhile sparse had been updated with a specialized implementation for `map` which was causing this PR to fail. For now, this PR has a naive implementation - should be optimized later. Added a couple more tests.

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,365 @@
+and collects output into results.
+
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+`[asyncmap()`](:func:`asyncmap`). If `batch_size` is specified, `f` must
```
I think it's supposed to be backticks inside square brackets

amitmurthy committed:
docstring changes

amitmurthy commented:
Rebased after https://github.com/JuliaLang/julia/pull/18588 . `make docs` is not capturing updated docstrings. Anything else to be done?

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,365 @@
+
+`ntasks` specifies the number of tasks to run concurrently.
+Depending on the length of the collections, if `ntasks` is unspecified,
+up to a 100 tasks will be used for concurrent mapping.
```
up to 100

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,365 @@
+    AsyncCollector(f, results, c...; ntasks=0, batch_size=nothing) -> iterator
+
+Returns an iterator which applies `f` to each element of `c` asynchronously
+and collects output into results.
```
code highlight `results`

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,365 @@
+!!! note
+    `next(::AsyncCollector, state) -> (nothing, state)`. A successful return
+    from `next` indicates that the next element from the input collection is
+    being processed asynchronously. It blocks till a free worker task becomes
```
blocks until

amitmurthy replied:
Indian English comes naturally to me. Will update.

tkelman commented on base/asyncmap.jl:
```diff
@@ -3,161 +3,365 @@
+
+!!! note
+    `for _ in AsyncCollector(f, results, c...) end` is equivalent to
+    `map!(f, results, c...)`.
```
is it equivalent to map! or asyncmap! ?

amitmurthy replied:
Both. The former in terms of returned data. The latter in terms of asynchronous execution.

amitmurthy replied:
Will update.

tkelman commented on base/asyncmap.jl:
```diff
@@ -220,26 +413,21 @@ size(itr::AsyncGenerator) = size(itr.collector.enumerator)
-    asyncmap!(f, results, c...)
+    asyncmap!(f, results, c...; ntasks=0, batch_size=nothing)
 
 Like [`asyncmap()`](@ref), but stores output in `results` rather returning a collection.
```
rather than returning

amitmurthy committed:
doc fixups

MichaelHatherly commented:
> Rebased after #18588 . `make docs` is not capturing updated docstrings. Anything else to be done?

Docs for `asyncmap` and `asyncmap!` appear to be showing up when I build this branch locally. Are there any error messages being printed?

amitmurthy commented:
Shows up in the REPL. Doesn't show up in `doc/_build/html/stdlib/parallel.html`. 
```
Documenter: setting up build directory.
Documenter: expanding markdown templates.
Documenter: building cross-references.
Documenter: running document checks.
 !! Skipped doctesting.
 > checking footnote links.
 !! Unused footnote named 'BH83'. [src/stdlib/linalg.md]
Documenter: populating indices.
Documenter: rendering document.
 !! Overwriting '_build/html/en/assets/documenter.css'.
 !! Overwriting '_build/html/en/assets/documenter.js'.
 !! Overwriting '_build/html/en/assets/search.js'.
Build finished. The HTML pages are in _build/html.
```

MichaelHatherly commented:
> `doc/_build/html/stdlib/parallel.html`

Should be `doc/_build/html/en/stdlib/parallel.html` instead. Perhaps that's just an older build in `doc/_build/html/stdlib/parallel.html`.

amitmurthy commented:
Right! Thanks.

Merging this PR in a short while.

andreasnoack commented on base/asyncmap.jl:
```diff
@@ -220,26 +413,22 @@ size(itr::AsyncGenerator) = size(itr.collector.enumerator)
-asyncmap!(f, r, c1, c...) = (for x in AsyncCollector(f, r, c1, c...) end; r)
+function asyncmap!(f, r, c1, c...; ntasks=0, batch_size=nothing)
+    foreach(identity, AsyncCollector(f, r, c1, c...; ntasks=ntasks, batch_size=batch_size))
+    c
```
@amitmurthy This should `r`, right?

amitmurthy replied:
Yes.
